File: go/Assembler/engine/engine_preprocess_json.go
package engine

import (
	Logger "arshu/common"
	"assembler/common"
	"assembler/model"
	"fmt"
	"strings"
)

type ILoaderPreprocessJson interface {
	GetTemplateHtml(appSite, appFile, appView, appViewPrefix string) *model.PreprocessedTemplate
	GetTemplateJson(appSite, appFile string) map[string]interface{}
	AllTemplates() map[string]*model.PreprocessedTemplate
}

type EnginePreProcessJson struct {
	AppViewPrefix string
}

func (e *EnginePreProcessJson) MergeTemplates(appSite, appFile, appView string, loader ILoaderPreprocessJson, enableJsonProcessing bool) string {
	Logger.Debug(fmt.Sprintf("MergeTemplates called: appSite=%s, appFile=%s, appView=%s, enableJson=%t", appSite, appFile, appView, enableJsonProcessing), "EnginePreProcessJson")

	if loader == nil {
		Logger.Debug("No loader provided", "EnginePreProcessJson")
		return ""
	}

	preprocessedTemplates := loader.AllTemplates()
	if len(preprocessedTemplates) == 0 {
		Logger.Debug("No preprocessed templates available", "EnginePreProcessJson")
		return ""
	}

	Logger.Debug(fmt.Sprintf("Using %d preprocessed templates", len(preprocessedTemplates)), "EnginePreProcessJson")

	mainPreprocessed := loader.GetTemplateHtml(appSite, appFile, appView, e.AppViewPrefix)
	if mainPreprocessed == nil {
		Logger.Debug(fmt.Sprintf("Main template not found for appSite=%s, appFile=%s", appSite, appFile), "EnginePreProcessJson")
		return ""
	}

	Logger.Debug(fmt.Sprintf("Main template found, original size: %d", len(mainPreprocessed.OriginalContent)), "EnginePreProcessJson")

	contentHtml := mainPreprocessed.OriginalContent

	if enableJsonProcessing && mainPreprocessed.JsonData != nil {
		Logger.Debug("Merging main template JSON", "EnginePreProcessJson")
		contentHtml = MergeTemplateWithJson(contentHtml, *mainPreprocessed.JsonData)
	}

	contentHtml = e.applyTemplateReplacements(contentHtml, preprocessedTemplates, enableJsonProcessing, appView, mainPreprocessed, loader, appSite)

	Logger.Debug(fmt.Sprintf("MergeTemplates complete: output size=%d", len(contentHtml)), "EnginePreProcessJson")
	return contentHtml
}

func (e *EnginePreProcessJson) getTemplate(appSite, templateName string, preprocessedTemplates map[string]*model.PreprocessedTemplate, appView, appViewPrefix string, useAppViewFallback bool) *model.PreprocessedTemplate {
	if len(preprocessedTemplates) == 0 {
		return nil
	}

	viewPrefix := appViewPrefix
	if viewPrefix == "" {
		viewPrefix = e.AppViewPrefix
	}

	if useAppViewFallback && appView != "" && viewPrefix != "" && strings.Contains(templateName, viewPrefix) {
		appKey := common.ReplaceCaseInsensitive(templateName, viewPrefix, appView)
		fallbackTemplateKey := strings.ToLower(appSite) + "_" + strings.ToLower(appKey)
		if fallbackTemplate, ok := preprocessedTemplates[fallbackTemplateKey]; ok {
			return fallbackTemplate
		}
	}

	primaryTemplateKey := strings.ToLower(appSite) + "_" + strings.ToLower(templateName)
	if primaryTemplate, ok := preprocessedTemplates[primaryTemplateKey]; ok {
		return primaryTemplate
	}

	return nil
}

func (e *EnginePreProcessJson) applyTemplateReplacements(content string, preprocessedTemplates map[string]*model.PreprocessedTemplate, enableJsonProcessing bool, appView string, mainTemplate *model.PreprocessedTemplate, loader ILoaderPreprocessJson, appSite string) string {
	result := content
	Logger.Debug(fmt.Sprintf("Starting ApplyTemplateReplacements, initial size: %d", len(content)), "EnginePreProcessJson")

	previous := ""
	maxPasses := 10
	currentPass := 0

	for {
		previous = result
		currentPass++

		Logger.Debug(fmt.Sprintf("Replacement pass %d, current size: %d", currentPass, len(result)), "EnginePreProcessJson")

		slottedCount, simpleCount, jsonPlaceholderCount := 0, 0, 0

		if mainTemplate != nil && currentPass == 1 && enableJsonProcessing {
			for _, mapping := range mainTemplate.ReplacementMappings {
				if mapping.Type != model.JsonPlaceholderType {
					continue
				}
				if strings.Contains(result, mapping.OriginalText) {
					Logger.Debug(fmt.Sprintf("Applying main template JSON placeholder: %s -> %s", mapping.OriginalText, mapping.ReplacementText), "EnginePreProcessJson")
					result = strings.ReplaceAll(result, mapping.OriginalText, mapping.ReplacementText)
					jsonPlaceholderCount++
				}
			}
		}

		for _, template := range preprocessedTemplates {
			for _, mapping := range template.ReplacementMappings {
				if mapping.Type != model.SlottedTemplateType {
					continue
				}
				if strings.Contains(result, mapping.OriginalText) {
					replacementText := mapping.ReplacementText
					if enableJsonProcessing && mapping.TargetTemplateName != "" {
						targetJson := loader.GetTemplateJson(appSite, mapping.TargetTemplateName)
						if targetJson != nil {
							Logger.Debug(fmt.Sprintf("Merging JSON for slotted template %s", mapping.TargetTemplateName), "EnginePreProcessJson")
							replacementText = MergeTemplateWithJson(replacementText, targetJson)
						}
					}
					Logger.Debug(fmt.Sprintf("Applying slotted template: %s... -> %d chars", mapping.OriginalText[:min(50, len(mapping.OriginalText))], len(replacementText)), "EnginePreProcessJson")
					result = strings.ReplaceAll(result, mapping.OriginalText, replacementText)
					slottedCount++
				}
			}

			for _, mapping := range template.ReplacementMappings {
				if mapping.Type != model.SimpleTemplateType {
					continue
				}
				if strings.Contains(result, mapping.OriginalText) {
					replacementText := mapping.ReplacementText
					if appView != "" && mapping.TargetTemplateName != "" {
						appViewTemplate := e.getTemplate(appSite, mapping.TargetTemplateName, preprocessedTemplates, appView, e.AppViewPrefix, true)
						if appViewTemplate != nil {
							replacementText = appViewTemplate.OriginalContent
						}
					}

					if enableJsonProcessing && mapping.TargetTemplateName != "" {
						targetJson := loader.GetTemplateJson(appSite, mapping.TargetTemplateName)
						if targetJson != nil {
							Logger.Debug(fmt.Sprintf("Merging JSON for simple template %s", mapping.TargetTemplateName), "EnginePreProcessJson")
							replacementText = MergeTemplateWithJson(replacementText, targetJson)
						}
					}
					Logger.Debug(fmt.Sprintf("Applying simple template: %s -> %d chars", mapping.OriginalText, len(replacementText)), "EnginePreProcessJson")
					result = strings.ReplaceAll(result, mapping.OriginalText, replacementText)
					simpleCount++
				}
			}
		}

		Logger.Debug(fmt.Sprintf("Pass %d applied: %d main JSON placeholders, %d slotted, %d simple", currentPass, jsonPlaceholderCount, slottedCount, simpleCount), "EnginePreProcessJson")

		if result == previous || currentPass >= maxPasses {
			break
		}
	}

	Logger.Debug(fmt.Sprintf("Replacement complete after %d passes, final size: %d", currentPass, len(result)), "EnginePreProcessJson")
	return result
}

File: php/Assembler/src/Engine/EnginePreProcessJson.php
<?php

namespace Assembler\Engine;

use Arshu\Common\Logger;
use Assembler\Common\CommonUtil;
use Assembler\Loader\LoaderPreProcessJson;
use Assembler\Model\ReplacementType;

class EnginePreProcessJson
{
    private string $appViewPrefix = '';

    public function setAppViewPrefix(string $prefix): void
    {
        $this->appViewPrefix = $prefix;
    }

    public function mergeTemplates(string $appSite, string $appFile, ?string $appView, LoaderPreProcessJson $loader, bool $enableJsonProcessing = true): string
    {
        Logger::debug("MergeTemplates called: appSite=$appSite, appFile=$appFile, appView=" . ($appView ?? 'null') . ", enableJson=$enableJsonProcessing", 'EnginePreProcessJson');

        if (!$loader) {
            Logger::warn('No loader provided', 'EnginePreProcessJson');
            return '';
        }

        $preprocessedTemplates = $loader->getAllTemplates();
        if (empty($preprocessedTemplates)) {
            Logger::warn('No preprocessed templates available', 'EnginePreProcessJson');
            return '';
        }

        Logger::debug('Using ' . count($preprocessedTemplates) . ' preprocessed templates', 'EnginePreProcessJson');

        $mainPreprocessed = $loader->getTemplateHtml($appSite, $appFile, $appView, $this->appViewPrefix);
        if (!$mainPreprocessed) {
            Logger::warn("Main template not found for appSite=$appSite, appFile=$appFile", 'EnginePreProcessJson');
            return '';
        }

        Logger::debug('Main template found, original size: ' . strlen($mainPreprocessed->getOriginalContent()), 'EnginePreProcessJson');

        $contentHtml = $mainPreprocessed->getOriginalContent();

        if ($enableJsonProcessing && $mainPreprocessed->getJsonData()) {
            Logger::debug('Merging main template JSON', 'EnginePreProcessJson');
            $contentHtml = JsonMergeUtil::mergeTemplateWithJson($contentHtml, $mainPreprocessed->getJsonData());
        }

        $contentHtml = $this->applyTemplateReplacements($contentHtml, $preprocessedTemplates, $enableJsonProcessing, $appView, $mainPreprocessed, $loader, $appSite);

        Logger::debug('MergeTemplates complete: output size=' . strlen($contentHtml), 'EnginePreProcessJson');
        return $contentHtml;
    }

    private function getTemplate(string $appSite, string $templateName, array $preprocessedTemplates, ?string $appView = null, ?string $appViewPrefix = null, bool $useAppViewFallback = true)
    {
        if (empty($preprocessedTemplates)) {
            return null;
        }

        $viewPrefix = $appViewPrefix ?? $this->appViewPrefix;

        if ($useAppViewFallback && $appView && $viewPrefix && stripos($templateName, $viewPrefix) !== false) {
            $appKey = CommonUtil::replaceCaseInsensitive($templateName, $viewPrefix, $appView);
            $fallbackTemplateKey = strtolower($appSite) . '_' . strtolower($appKey);
            if (isset($preprocessedTemplates[$fallbackTemplateKey])) {
                return $preprocessedTemplates[$fallbackTemplateKey];
            }
        }

        $primaryTemplateKey = strtolower($appSite) . '_' . strtolower($templateName);
        if (isset($preprocessedTemplates[$primaryTemplateKey])) {
            return $preprocessedTemplates[$primaryTemplateKey];
        }

        return null;
    }

    private function applyTemplateReplacements(string $content, array $preprocessedTemplates, bool $enableJsonProcessing, ?string $appView, $mainTemplate, LoaderPreProcessJson $loader, string $appSite): string
    {
        $result = $content;
        Logger::debug('Starting ApplyTemplateReplacements, initial size: ' . strlen($content), 'EnginePreProcessJson');

        $previous = '';
        $maxPasses = 10;
        $currentPass = 0;

        do {
            $previous = $result;
            $currentPass++;
            Logger::debug("Replacement pass $currentPass, current size: " . strlen($result), 'EnginePreProcessJson');

            $slottedCount = 0;
            $simpleCount = 0;
            $jsonPlaceholderCount = 0;

            if ($mainTemplate && $currentPass === 1 && $enableJsonProcessing) {
                foreach ($mainTemplate->getReplacementMappings() as $i => $mapping) {
                    Logger::debug("[Main JSON Placeholder] Mapping #$i: " . json_encode($mapping), 'EnginePreProcessJson');
                    if (method_exists($mapping, 'getType') && $mapping->getType() !== ReplacementType::JSON_PLACEHOLDER) continue;
                    if (method_exists($mapping, 'getOriginalText') && strpos($result, $mapping->getOriginalText()) !== false) {
                        Logger::debug('Applying main template JSON placeholder: ' . $mapping->getOriginalText() . ' -> ' . $mapping->getReplacementText(), 'EnginePreProcessJson');
                        $result = str_replace($mapping->getOriginalText(), $mapping->getReplacementText(), $result);
                        $jsonPlaceholderCount++;
                    }
                }
            }

            foreach ($preprocessedTemplates as $tidx => $template) {
                Logger::debug("[Template #$tidx] Processing template: " . (method_exists($template, 'getOriginalContent') ? $template->getOriginalContent() : 'N/A'), 'EnginePreProcessJson');
                foreach ($template->getReplacementMappings() as $mid => $mapping) {
                    Logger::debug("[Slotted/Simple] Mapping #$mid: " . json_encode($mapping), 'EnginePreProcessJson');
                    if (method_exists($mapping, 'getType') && $mapping->getType() === ReplacementType::SLOTTED_TEMPLATE) {
                        if (method_exists($mapping, 'getOriginalText') && strpos($result, $mapping->getOriginalText()) !== false) {
                            $replacementText = $mapping->getReplacementText();
                            if ($enableJsonProcessing && method_exists($mapping, 'getTargetTemplateName') && $mapping->getTargetTemplateName()) {
                                $targetJson = $loader->getTemplateJson($appSite, $mapping->getTargetTemplateName());
                                if ($targetJson) {
                                    Logger::debug('Merging JSON for slotted template ' . $mapping->getTargetTemplateName(), 'EnginePreProcessJson');
                                    $replacementText = JsonMergeUtil::mergeTemplateWithJson($replacementText, $targetJson);
                                }
                            }
                            Logger::debug('Applying slotted template: ' . substr($mapping->getOriginalText(), 0, 50) . '... -> ' . strlen($replacementText) . ' chars', 'EnginePreProcessJson');
                            $result = str_replace($mapping->getOriginalText(), $replacementText, $result);
                            $slottedCount++;
                        }
                    } elseif (method_exists($mapping, 'getType') && $mapping->getType() === ReplacementType::SIMPLE_TEMPLATE) {
                        if (method_exists($mapping, 'getOriginalText') && strpos($result, $mapping->getOriginalText()) !== false) {
                            $replacementText = $mapping->getReplacementText();
                            if ($appView && method_exists($mapping, 'getTargetTemplateName') && $mapping->getTargetTemplateName()) {
                                $appViewTemplate = $this->getTemplate($appSite, $mapping->getTargetTemplateName(), $preprocessedTemplates, $appView, $this->appViewPrefix, true);
                                if ($appViewTemplate) {
                                    $replacementText = $appViewTemplate->getOriginalContent();
                                }
                            }

                            if ($enableJsonProcessing && method_exists($mapping, 'getTargetTemplateName') && $mapping->getTargetTemplateName()) {
                                $targetJson = $loader->getTemplateJson($appSite, $mapping->getTargetTemplateName());
                                if ($targetJson) {
                                    Logger::debug('Merging JSON for simple template ' . $mapping->getTargetTemplateName(), 'EnginePreProcessJson');
                                    $replacementText = JsonMergeUtil::mergeTemplateWithJson($replacementText, $targetJson);
                                }
                            }
                            Logger::debug('Applying simple template: ' . $mapping->getOriginalText() . ' -> ' . strlen($replacementText) . ' chars', 'EnginePreProcessJson');
                            $result = str_replace($mapping->getOriginalText(), $replacementText, $result);
                            $simpleCount++;
                        }
                    }
                }
            }
            Logger::debug("Pass $currentPass applied: $jsonPlaceholderCount main JSON placeholders, $slottedCount slotted, $simpleCount simple", 'EnginePreProcessJson');
        } while ($result !== $previous && $currentPass < $maxPasses);

        Logger::debug("Replacement complete after $currentPass passes, final size: " . strlen($result), 'EnginePreProcessJson');
        return $result;
    }
}